File: Sources/Syrtis/DisabledProviders.swift
import Foundation
import TokenBarCore

/// The quota providers the user has switched off in Settings, and the seam that
/// hands that list to the FFI registry.
///
/// **This is not a card-visibility setting.** Hiding a card would leave the
/// fetch running, and the fetch is the expensive part: `tb_agent_usage` joins
/// every provider and returns only when the slowest finishes, so one slow
/// provider delays every other card. The Antigravity CLI fallback is the case
/// that made this worth a setting — with no Antigravity IDE installed it spawns
/// a ~170MB binary on every publication, measured at 2.6-4.6s, and the Claude
/// limits card could not appear until that returned. Switching a provider off
/// here stops the future being created at all.
///
/// Stored as a comma-separated raw string, the same shape
/// `tokenbar.views.hidden` uses, so a `@AppStorage` binding in Settings is the
/// whole UI contract.
enum DisabledProviders {
    static let storageKey = "tokenbar.providers.disabled"

    /// Every provider the core can fetch, in the order Settings lists them.
    /// Mirrors `disabled_providers::KNOWN_PROVIDERS`; an id absent from the
    /// Rust list is refused by the setter rather than silently stored.
    static let known: [String] = [
        "claude", "codex", "antigravity", "copilot", "grok", "grok-bot", "kiro", "opencode-go",
    ]

    /// Display name for a provider row. Deliberately the vendor's own casing
    /// rather than a localized string: these are product names.
    static func label(_ id: String) -> String {
        switch id {
        case "claude": return "Claude"
        case "codex": return "Codex"
        case "antigravity": return "Antigravity"
        case "copilot": return "GitHub Copilot"
        case "grok": return "Grok"
        case "grok-bot": return "Grok Bot"
        case "kiro": return "Kiro"
        case "opencode-go": return "OpenCode Go"
        default: return id
        }
    }

    /// Parse the stored raw value. Unknown ids are dropped here as well as in
    /// Rust: a value written by a newer build that knew a provider this one
    /// does not must not disable something at random.
    static func parse(raw: String) -> [String] {
        var seen = Set<String>()
        return raw.split(separator: ",")
            .map { $0.trimmingCharacters(in: .whitespaces) }
            .filter { known.contains($0) && seen.insert($0).inserted }
    }

    static func current() -> [String] {
        parse(raw: UserDefaults.standard.string(forKey: storageKey) ?? "")
    }

    /// The exact wire shape the setter expects: a JSON array of ids. An empty
    /// selection is an explicit `[]` rather than an omitted call, because the
    /// setter is full-replace and `[]` is how a provider gets re-enabled.
    static func payloadJSON(_ ids: [String]) -> String {
        guard let data = try? JSONSerialization.data(withJSONObject: ids),
              let json = String(data: data, encoding: .utf8)
        else { return "[]" }
        return json
    }

    /// Last value this process installed. The Rust registry starts empty every
    /// launch, so a persisted marker cannot answer "is the core already holding
    /// this"; only an in-process record can.
    @MainActor private static var lastInstalled: String?

    /// Push the stored selection into the core. Safe to call on every launch
    /// and on every Settings write.
    ///
    /// Failure is swallowed on purpose, matching `ClaudeExtraRoots.apply`: a
    /// refused registry install must not take down launch, and the consequence
    /// is a provider that keeps being fetched — slower, never wrong.
    @MainActor
    static func apply() {
        let ids = current()
        let json = payloadJSON(ids)
        let isFirstInstall = lastInstalled == nil
        let changed = lastInstalled != json
        lastInstalled = json
        _ = try? TBCore.setDisabledProviders(json: json)
        // A first install of an empty selection writes the state the registry
        // already had, so nothing needs waking. Without this, every launch
        // would invalidate the throttle and restart the poll for no change.
        guard changed, !(isFirstInstall && ids.isEmpty) else { return }
        Task { @MainActor in
            // Same order and same reason as the account registry: drop the
            // throttled payload BEFORE waking the poll, or the woken fetch is
            // answered from a payload built for the provider set that just
            // changed — and the user waits out the 50s floor watching a card
            // they just switched off.
            await AgentUsageThrottle.shared.invalidate()
            ClaudeExtraRoots.RegistryChange.signal()
        }
    }

    /// Test seam: install an explicit selection without touching UserDefaults
    /// or the in-process record used by `apply`.
    static func installForTesting(_ ids: [String]) throws -> DisabledProvidersResult {
        try TBCore.setDisabledProviders(json: payloadJSON(ids))
    }

    /// Test seam: forget what this process installed, so a suite can drive
    /// `apply`'s first-install branch more than once.
    @MainActor
    static func resetInstalledRecordForTesting() {
        lastInstalled = nil
    }
}

File: crates/tb_core_ffi/src/disabled_providers.rs
//! Process-wide registry of the quota providers the user has switched off in
//! Settings. Written by `tb_set_disabled_providers`, read by `agent_usage::run`
//! before it builds the provider futures.
//!
//! **Why a registry and not a Swift-side filter.** Hiding a card in the UI is a
//! presentation decision; it does not stop the fetch, and the fetch is the part
//! that costs. One provider's fetch can hold up every other card, because
//! `run` joins them and returns only when the slowest finishes: the Antigravity
//! CLI fallback spawns a ~170MB binary that takes seconds, so a user who does
//! not use Antigravity was paying that on every publication for a card they did
//! not want. A filter that runs after the payload is built cannot recover that
//! time. This one runs before the future is created, so a disabled provider
//! costs nothing at all.
//!
//! A `RwLock` static rather than an env var, for the same reason
//! `extra_scan_paths` and `claude_config_dirs` are: the process is resident and
//! a Settings edit must take effect without a restart.
//!
//! The registry starts empty every launch — it is in-memory, not persisted here
//! — so Swift owns re-applying it from `UserDefaults` at launch and after every
//! edit. A process that never calls the setter fetches exactly the providers it
//! fetched before this module existed.

use std::collections::BTreeSet;
use std::sync::{LazyLock, RwLock};

/// Every provider id `agent_usage::run` can fetch. An id outside this set is
/// refused rather than stored: it would sit in the registry disabling nothing,
/// and a typo in a Settings value would look exactly like a working toggle.
pub(crate) const KNOWN_PROVIDERS: [&str; 8] = [
    "codex",
    "claude",
    "antigravity",
    "copilot",
    "grok",
    "grok-bot",
    "kiro",
    "opencode-go",
];

static DISABLED_PROVIDERS: LazyLock<RwLock<BTreeSet<String>>> =
    LazyLock::new(|| RwLock::new(BTreeSet::new()));

/// The provider ids currently switched off. Empty by default, which is the
/// pre-feature behavior exactly.
pub(crate) fn snapshot() -> BTreeSet<String> {
    DISABLED_PROVIDERS
        .read()
        .unwrap_or_else(|poisoned| poisoned.into_inner())
        .clone()
}

/// Replace the whole registry from a JSON array of provider ids.
/// Full-replace, not merge: `[]` re-enables everything (the Settings rollback
/// path).
///
/// A repeated id is not an error the way a repeated Claude config directory is.
/// That one produces two cards writing one series; this one is a set, and
/// listing `claude` twice asks for the same state as listing it once. It is
/// reported as accepted and folded into the set.
pub(crate) fn set_from_json(raw: &str) -> Result<serde_json::Value, String> {
    let input: Vec<String> = serde_json::from_str(raw)
        .map_err(|e| format!("invalid disabled providers JSON: {}", e))?;

    let mut disabled: BTreeSet<String> = BTreeSet::new();
    let mut rejected: Vec<serde_json::Value> = Vec::new();
    for raw_id in input {
        if KNOWN_PROVIDERS.contains(&raw_id.as_str()) {
            disabled.insert(raw_id);
        } else {
            rejected.push(serde_json::json!({
                "id": raw_id,
                "reason": "unknown provider id",
            }));
        }
    }

    let disabled_count = disabled.len();
    *DISABLED_PROVIDERS
        .write()
        .unwrap_or_else(|poisoned| poisoned.into_inner()) = disabled;

    Ok(serde_json::json!({
        "disabledCount": disabled_count,
        "rejected": rejected,
    }))
}

/// One process-wide mutex for every test that reads or writes the static, so
/// parallel `cargo test` threads do not observe each other's registry.
#[cfg(test)]
pub(crate) static TEST_LOCK: std::sync::Mutex<()> = std::sync::Mutex::new(());

#[cfg(test)]
pub(crate) fn reset_for_test() {
    *DISABLED_PROVIDERS
        .write()
        .unwrap_or_else(|poisoned| poisoned.into_inner()) = BTreeSet::new();
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn registers_known_ids_and_reports_what_it_refused() {
        let _guard = TEST_LOCK.lock().unwrap_or_else(|p| p.into_inner());
        reset_for_test();

        let result =
            set_from_json(r#"["antigravity", "grok", "Antigravity", "", "antigravity"]"#).unwrap();

        // Two distinct ids: the repeat folds into the set, the wrong-case and
        // empty ids are refused.
        assert_eq!(result["disabledCount"], 2);
        assert_eq!(
            snapshot(),
            BTreeSet::from(["antigravity".to_string(), "grok".to_string()])
        );
        let reasons: Vec<&str> = result["rejected"]
            .as_array()
            .unwrap()
            .iter()
            .map(|note| note["reason"].as_str().unwrap())
            .collect();
        assert_eq!(reasons, vec!["unknown provider id", "unknown provider id"]);

        reset_for_test();
    }

    #[test]
    fn full_replace_re_enables_everything() {
        let _guard = TEST_LOCK.lock().unwrap_or_else(|p| p.into_inner());
        reset_for_test();

        set_from_json(r#"["antigravity"]"#).unwrap();
        assert!(!snapshot().is_empty());
        set_from_json("[]").unwrap();
        assert!(
            snapshot().is_empty(),
            "an empty array is the Settings rollback path and must re-enable \
             every provider, not merge into the previous set"
        );

        reset_for_test();
    }

    #[test]
    fn malformed_json_leaves_the_registry_untouched() {
        let _guard = TEST_LOCK.lock().unwrap_or_else(|p| p.into_inner());
        reset_for_test();
        set_from_json(r#"["antigravity"]"#).unwrap();
        let error = set_from_json("{not json").unwrap_err();
        assert!(error.contains("invalid disabled providers JSON"), "{error}");
        assert_eq!(snapshot(), BTreeSet::from(["antigravity".to_string()]));
        reset_for_test();
    }

    #[test]
    fn every_known_provider_id_is_one_run_actually_fetches() {
        // The set is duplicated from `agent_usage::run`'s join by necessity —
        // the futures there are named, not enumerable. This pins the copy so a
        // provider renamed on one side cannot leave a toggle that silently
        // disables nothing.
        assert_eq!(
            KNOWN_PROVIDERS.len(),
            8,
            "a provider was added or removed in run(); update this list and the \
             Settings UI, or the new provider gets no toggle"
        );
    }
}
